virtcontainers/cgroups.go

```diff
@@ -1,197 +1,251 @@
 	}
 
-	// apply cpu constraint to vcpu cgroup
-	if err := s.cgroup.vcpuSub.Update(rc); err != nil {
-		return err
+	return noConstraintsCgroup.Delete()
+}
+
+func (s *Sandbox) constrainHypervisor(cgroup cgroups.Cgroup) error {
+	pid := s.hypervisor.pid()
+	if pid <= 0 {
+		return fmt.Errorf("Invalid hypervisor PID: %d", pid)
 	}
 
-	// when new container joins, new CPU could be hotplugged, so we
-	// have to query fresh vcpu info from hypervisor for every time.
-	tids, err := s.hypervisor.getThreadIDs()
+	resources := &specs.LinuxResources{}
+	path := cgroupNoConstraintsPath(s.state.CgroupPath)
+	noConstraintsCgroup, err := cgroups.New(V1NoConstraints, cgroups.StaticPath(path), resources)
 	if err != nil {
-		return fmt.Errorf("failed to get thread ids from hypervisor: %v", err)
-	}
-	if tids == nil {
-		// If there's no tid returned from the hypervisor, this is not
-		// a bug. It simply means there is nothing to constrain, hence
-		// let's return without any error from here.
-		return nil
+		return fmt.Errorf("Could not create cgroup %v: %v", path, err)
 	}
 
-	// use Add() to add vcpu thread to s.cgroup, it will write thread id to
-	// `cgroup.procs` which will move all threads in qemu process to this cgroup
-	// immediately as default behaviour.
-	if len(tids.vcpus) > 0 {
-		if err := s.cgroup.sandboxSub.Add(cgroups.Process{
-			Pid: tids.vcpus[0],
-		}); err != nil {
-			return err
-		}
+	if err := noConstraintsCgroup.Add(cgroups.Process{Pid: pid}); err != nil {
+		return fmt.Errorf("Could not add hypervisor PID %d to cgroup %v: %v", pid, path, err)
 	}
 
-	for _, i := range tids.vcpus {
-		if i <= 0 {
-			continue
-		}
+	// Add qemu into cgroup
+	return cgroup.Add(cgroups.Process{Pid: pid})
+}
 
-		// In contrast, AddTask will write thread id to `tasks`
-		// After this, vcpu threads are in "vcpu" sub-cgroup, other threads in
-		// qemu will be left in parent cgroup untouched.
-		if err := s.cgroup.vcpuSub.AddTask(cgroups.Process{
-			Pid: i,
-		}); err != nil {
-			return err
-		}
+func (s *Sandbox) resources() (specs.LinuxResources, error) {
+	resources := specs.LinuxResources{
+		CPU: s.cpuResources(),
 	}
 
-	return nil
+	return resources, nil
 }
 
-func (s *Sandbox) mergeSpecResource() (*specs.LinuxResources, error) {
-	if s.config == nil {
-		return nil, fmt.Errorf("sandbox config is nil")
+func (s *Sandbox) cpuResources() *specs.LinuxCPU {
+	quota := int64(0)
+	period := uint64(0)
+	shares := uint64(0)
+	cpus := ""
+
+	for _, c := range s.containers {
+		if s.state.Pid != c.process.Pid {
+			if c.state.Resources.CPU == nil {
+				continue
+			}
+
+			if c.state.Resources.CPU.Shares != nil {
+				shares = uint64(math.Max(float64(*c.state.Resources.CPU.Shares), float64(shares)))
+			}
+
+			if c.state.Resources.CPU.Quota != nil {
+				quota += *c.state.Resources.CPU.Quota
+			}
+
+			if c.state.Resources.CPU.Period != nil {
+				period = uint64(math.Max(float64(*c.state.Resources.CPU.Period), float64(period)))
+			}
+
```

Comment:
Missing `RealtimePeriod` and `RealtimeRuntime` ?

Reply:
fixed